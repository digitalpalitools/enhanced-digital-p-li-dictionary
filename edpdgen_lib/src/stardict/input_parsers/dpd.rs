use crate::inflection_generator::InflectionGenerator;
use crate::stardict::input_parsers::{make_group_id, make_sort_key, make_toc_id, PaliWord};
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "dpd_toc_summary",
            include_str!("templates/dpd_toc_summary.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.add_raw_templates(vec![(
            "dpd_word_data",
            include_str!("templates/dpd_word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
}

// NOTE: Keep the order deliberately randomized as we support column reordering.
#[derive(Debug, Serialize, Deserialize)]
pub struct DpdPaliWord {
    #[serde(rename = "Pāli1")]
    pali1: String,
    #[serde(rename = "Pāli2")]
    pali2: String,
    #[serde(rename = "Fin")]
    fin: String,
    #[serde(rename = "POS")]
    pos: String,
    #[serde(rename = "Grammar")]
    grammar: String,
    #[serde(rename = "Derived from")]
    derived_from: String,
    #[serde(rename = "Neg")]
    neg: String,
    #[serde(rename = "Verb")]
    verb: String,
    #[serde(rename = "Trans")]
    trans: String,
    #[serde(rename = "Case")]
    case: String,
    #[serde(rename = "Meaning IN CONTEXT")]
    in_english: String,
    #[serde(rename = "Sanskrit")]
    sanskrit: String,
    #[serde(rename = "Sk Root")]
    sanskrit_root: String,
    #[serde(rename = "Family")]
    family: String,
    #[serde(rename = "Pāli Root")]
    pali_root: String,
    #[serde(rename = "V")]
    v: String,
    #[serde(rename = "Grp")]
    grp: String,
    #[serde(rename = "Sgn")]
    sgn: String,
    #[serde(rename = "Root Meaning")]
    root_meaning: String,
    #[serde(rename = "Base")]
    base: String,
    #[serde(rename = "Construction")]
    construction: String,
    #[serde(rename = "Derivative")]
    derivative: String,
    #[serde(rename = "Suffix")]
    suffix: String,
    #[serde(rename = "Compound")]
    compound: String,
    #[serde(rename = "Compound Construction")]
    compound_construction: String,
    #[serde(rename = "Source1")]
    source1: String,
    #[serde(rename = "Sutta1")]
    sutta1: String,
    #[serde(rename = "Example1")]
    example1: String,
    #[serde(rename = "Source 2")]
    source2: String,
    #[serde(rename = "Sutta2")]
    sutta2: String,
    #[serde(rename = "Example 2")]
    example2: String,
    #[serde(rename = "Antonyms")]
    antonyms: String,
    #[serde(rename = "Synonyms – different word")]
    synonyms: String,
    #[serde(rename = "Variant – same constr or diff reading")]
    variant: String,
    #[serde(rename = "Commentary")]
    commentary: String,
    #[serde(rename = "Literal Meaning")]
    literal_meaning: String,
    #[serde(rename = "Root In Comps")]
    root_in_compound: String,
    #[serde(rename = "Notes")]
    notes: String,
    #[serde(rename = "Stem")]
    stem: String,
    #[serde(rename = "Pattern")]
    pattern: String,
    #[serde(rename = "Buddhadatta")]
    buddhadatta: String,
    #[serde(rename = "Cl")]
    cl: String,
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpdPaliWord,
    toc_id: &'a str,
    short_name: &'a str,
    feedback_form_url: &'a str,
    host_url: &'a str,
    host_version: &'a str,
    inflection_table: &'a str,
}

impl PaliWord for DpdPaliWord {
    fn id(&self) -> &str {
        &self.pali1
    }

    fn sort_key(&self) -> String {
        make_sort_key(&self.id())
    }

    fn group_id(&self) -> String {
        make_group_id(&self.id())
    }

    fn toc_id(&self, short_name: &str) -> String {
        make_toc_id(&self.id(), short_name)
    }

    fn toc_entry(&self, short_name: &str) -> Result<String, String> {
        let mut context = Context::new();
        context.insert("short_name", short_name);
        context.insert("toc_id", &self.toc_id(short_name));
        context.insert("pali1", &self.pali1);
        context.insert("pos", &self.pos);
        context.insert("in_english", &self.in_english);
        context.insert("buddhadatta", &self.buddhadatta);

        TEMPLATES
            .render("dpd_toc_summary", &context)
            .map_err(|e| e.to_string())
    }

    fn word_data_entry(
        &self,
        short_name: &str,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
    ) -> Result<String, String> {
        let vm = WordDataViewModel {
            word: &self,
            toc_id: &self.toc_id(short_name),
            short_name,
            feedback_form_url,
            host_url,
            host_version,
            inflection_table: &igen.generate_inflection_table_html(&self.pali1),
        };

        let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
        TEMPLATES
            .render("dpd_word_data", &context)
            .map_err(|e| e.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::stardict::input_parsers::load_words;
    use crate::stardict::input_parsers::tests::get_csv_path;
    use crate::tests::{TestInflectionGenerator, TestLogger};
    use test_case::test_case;

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn toc_summary_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let toc_summary = recs
            .nth(rec_number)
            .map(|r| r.toc_entry("dpy").expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(toc_summary);
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn word_data_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");
        let igen = TestInflectionGenerator::new();

        let word_data = recs
            .nth(rec_number)
            .map(|r| {
                r.word_data_entry("dpd", "fb_url", "host url", "host version", &igen)
                    .expect("unexpected")
            })
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
