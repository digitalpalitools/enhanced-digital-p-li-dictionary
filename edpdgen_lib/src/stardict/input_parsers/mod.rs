use crate::inflection_generator::InflectionGenerator;
use crate::{EdpdLogger, InputFormat};
use regex::{Captures, Regex};
use std::path::Path;

pub mod dpd;
pub mod dps;

lazy_static! {
    static ref PALI1_CRACKER: Regex = Regex::new(r"(.*)( )(\d+)$").expect("Malformed regex string");
}

pub trait PaliWord {
    fn id(&self) -> &str;
    fn sort_key(&self) -> String;
    fn group_id(&self) -> String;
    fn toc_id(&self, input_format: &InputFormat) -> String;
    fn toc_entry(&self, input_format: &InputFormat) -> Result<String, String>;
    fn word_data_entry(
        &self,
        input_format: &InputFormat,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
    ) -> Result<String, String>;
}

pub fn load_words<'a, T: 'a + serde::de::DeserializeOwned + PaliWord>(
    path: &Path,
    logger: &'a dyn EdpdLogger,
) -> Result<impl Iterator<Item = impl PaliWord> + 'a, String> {
    logger.info(&format!("Loading words from {:?}.", path));

    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
    let rdr = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .flexible(true)
        .from_reader(file);

    let words = rdr
        .into_deserialize::<T>()
        .enumerate()
        .filter_map(move |(i, r)| match r {
            Ok(w) => Some(w),
            Err(e) => {
                logger.error(&format!(
                    "Unable to deserialize record #{}. Error: {}.",
                    i, e
                ));
                None
            }
        });

    logger.info(&format!("... done loading words from {:?}.", &path));
    Ok(words)
}

fn make_sort_key(id: &str) -> String {
    let sk = PALI1_CRACKER.replace(id, |caps: &Captures| {
        // NOTE: Best case effort. Not sweating it.
        let n = &caps[3].parse::<i32>().unwrap_or(0);
        format!("{} {:03}", &caps[1], n)
    });

    sk.into_owned()
}

fn make_group_id(id: &str) -> String {
    let gid = PALI1_CRACKER.replace(id, |caps: &Captures| caps[1].to_string());

    gid.into_owned()
}

fn make_toc_id(id: &str, input_format: &InputFormat) -> String {
    format!("{}_{}", id.replace(" ", "_"), input_format.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::resolve_file_in_manifest_dir;
    use crate::stardict::input_parsers::dpd::DpdPaliWord;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0, "ābādha"; "0 digits")]
    #[test_case(9, "adhikāra 001"; "1 digit")]
    #[test_case(11, "adhikāra 010"; "2 digits")]
    fn test_sort_key(rec_number: usize, expected_sk: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("failed to load");

        let sk = recs
            .nth(rec_number)
            .map(|r| make_sort_key(r.id()))
            .expect("unexpected");

        assert_eq!(sk, expected_sk)
    }

    #[test_case(4, "abahulīkata"; "0 digits")]
    #[test_case(5, "abala"; "1 digit")]
    #[test_case(11, "adhikāra"; "2 digits")]
    fn test_group_id(rec_number: usize, expected_gid: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("failed to load");

        let gid = recs
            .nth(rec_number)
            .map(|r| make_group_id(r.id()))
            .expect("unexpected");

        assert_eq!(gid, expected_gid);
    }

    #[test_case(4, "abahulīkata_dps"; "0 digits")]
    #[test_case(5, "abala_1_dps"; "1 digit")]
    #[test_case(11, "adhikāra_10_dps"; "2 digits")]
    fn test_toc_id(rec_number: usize, expected_toc_id: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let toc_id = recs
            .nth(rec_number)
            .map(|r| make_toc_id(r.id(), &InputFormat::DevamittaPaliStudy))
            .expect("unexpected");

        assert_eq!(toc_id, expected_toc_id);
    }
}
