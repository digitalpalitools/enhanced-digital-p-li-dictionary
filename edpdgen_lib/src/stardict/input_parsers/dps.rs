use crate::inflection_generator::InflectionGenerator;
use crate::stardict::input_parsers::{make_group_id, make_sort_key, make_toc_id, PaliWord};
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "dps_toc_summary",
            include_str!("templates/dps_toc_summary.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.add_raw_templates(vec![(
            "dps_word_data",
            include_str!("templates/dps_word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
}

// NOTE: Keep the order deliberately randomized as we support column reordering.
#[derive(Debug, Serialize, Deserialize)]
pub struct DpsPaliWord {
    #[serde(rename = "Pāli1")]
    pali: String,
    #[serde(rename = "Fin")]
    fin: String,
    #[serde(rename = "POS")]
    pos: String,
    #[serde(rename = "Grammar")]
    grammar: String,
    #[serde(rename = "Derived from")]
    derived_from: String,
    #[serde(rename = "Neg")]
    neg: String,
    #[serde(rename = "Verb")]
    verb: String,
    #[serde(rename = "Trans")]
    trans: String,
    #[serde(rename = "Case")]
    case: String,
    #[serde(rename = "Meaning IN CONTEXT")]
    in_english: String,
    #[serde(rename = "Meaning in native language")]
    in_russian: String,
    #[serde(rename = "Pāli Root")]
    pali_root: String,
    #[serde(rename = "Base")]
    base: String,
    #[serde(rename = "Construction")]
    construction: String,
    #[serde(rename = "Sanskrit")]
    sanskrit: String,
    #[serde(rename = "Sk Root")]
    sanskrit_root: String,
    #[serde(rename = "Commentary")]
    commentary: String,
    #[serde(rename = "Notes")]
    notes: String,
    #[serde(rename = "Source1")]
    source1: String,
    #[serde(rename = "Example1")]
    example1: String,
    #[serde(rename = "Sutta1")]
    sutta1: String,
    #[serde(rename = "Source 2")]
    source2: String,
    #[serde(rename = "Example 2")]
    example2: String,
    #[serde(rename = "Sutta2")]
    sutta2: String,
    #[serde(rename = "Chapter")]
    chapter: String,
    #[serde(rename = "Test")]
    test: String,
    #[serde(rename = "Variant")]
    variant: String,
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpsPaliWord,
    toc_id: &'a str,
    short_name: &'a str,
    feedback_form_url: &'a str,
    host_url: &'a str,
    host_version: &'a str,
    inflection_table: &'a str,
}

impl PaliWord for DpsPaliWord {
    fn id(&self) -> &str {
        &self.pali
    }

    fn sort_key(&self) -> String {
        make_sort_key(&self.id())
    }

    fn group_id(&self) -> String {
        make_group_id(&self.id())
    }

    fn toc_id(&self, short_name: &str) -> String {
        make_toc_id(&self.id(), short_name)
    }

    fn toc_entry(&self, short_name: &str) -> Result<String, String> {
        let mut context = Context::new();
        context.insert("toc_id", &self.toc_id(short_name));
        context.insert("pali", &self.pali);
        context.insert("pos", &self.pos);
        context.insert("in_english", &self.in_english);

        TEMPLATES
            .render("dps_toc_summary", &context)
            .map_err(|e| e.to_string())
    }

    fn word_data_entry(
        &self,
        short_name: &str,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
    ) -> Result<String, String> {
        let vm = WordDataViewModel {
            word: &self,
            toc_id: &self.toc_id(short_name),
            short_name,
            feedback_form_url,
            host_url,
            host_version,
            inflection_table: &igen.generate_inflection_table_html(&self.pali),
        };

        let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
        TEMPLATES
            .render("dps_word_data", &context)
            .map_err(|e| e.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::resolve_file_in_manifest_dir;
    use crate::stardict::input_parsers::load_words;
    use crate::tests::{TestInflectionGenerator, TestLogger};
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("dps_sample.csv").expect("must exist!")
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    fn toc_summary_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpsPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let toc_summary = recs
            .nth(rec_number)
            .map(|r| r.toc_entry("dpy").expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(toc_summary);
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    fn word_data_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpsPaliWord>(&get_csv_path(), &l).expect("unexpected");
        let igen = TestInflectionGenerator::new();

        let word_data = recs
            .nth(rec_number)
            .map(|r| {
                r.word_data_entry("dpy", "fb_url", "host url", "host version", &igen)
                    .expect("unexpected")
            })
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
