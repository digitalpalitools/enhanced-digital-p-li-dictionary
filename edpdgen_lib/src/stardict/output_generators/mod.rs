use crate::inflection_generator::InflectionGenerator;
use crate::stardict::input_parsers::PaliWord;
use crate::stardict::{glib, StarDictFile};
use crate::{DictionaryInfo, EdpdLogger};
use itertools::Itertools;
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "word_group",
            include_str!("templates/word_group.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.add_raw_templates(vec![("ifo_file", include_str!("templates/ifo_file.txt"))])
            .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
}

#[derive(Debug, Serialize)]
struct IdxEntry {
    word: String,
    data_offset: i32,
    data_size: i32,
    synonym_words: Vec<String>,
}

#[derive(Serialize)]
struct WordGroupViewModel<'a> {
    ods_type: &'a str,
    accent_color: &'a str,
    toc_entries: &'a [String],
    descriptions: &'a [String],
}

#[derive(Serialize)]
struct IfoViewModel<'a> {
    name: &'a str,
    word_count: usize,
    syn_word_count: usize,
    idx_file_size: usize,
    author: &'a str,
    description: &'a str,
    time_stamp: &'a str,
}

fn log_return_error(
    w: &dyn PaliWord,
    short_msg: &str,
    e: String,
    logger: &dyn EdpdLogger,
) -> String {
    logger.warning(&format!(
        "Failed to generate {} for '{}'. Error: {}.",
        short_msg,
        w.id(),
        e
    ));
    e
}

fn get_ids_and_html_for_word_group(
    dict_info: &DictionaryInfo,
    words: impl Iterator<Item = impl PaliWord>,
    igen: &dyn InflectionGenerator,
    logger: &dyn EdpdLogger,
) -> Result<(Vec<String>, String), String> {
    let mut word_info: Vec<(String, String, String, String)> = words
        .map(|w| {
            (
                w.sort_key(),
                w.id().to_string(),
                w.toc_entry(&dict_info.input_format)
                    .unwrap_or_else(|e| log_return_error(&w, "table of contents", e, logger)),
                w.word_data_entry(
                    &dict_info.input_format,
                    dict_info.feedback_form_url,
                    dict_info.host_url,
                    dict_info.host_version,
                    igen,
                )
                .unwrap_or_else(|e| log_return_error(&w, "word data", e, logger)),
            )
        })
        .collect();
    word_info.sort_by(|a, b| a.0.cmp(&b.0));

    let (ids, toc_entries, descriptions) =
        word_info
            .into_iter()
            .fold((Vec::new(), Vec::new(), Vec::new()), |mut acc, e| {
                acc.0.push(e.1);
                acc.1.push(e.2);
                acc.2.push(e.3);
                acc
            });

    let vm = WordGroupViewModel {
        ods_type: &dict_info.input_format.to_string(),
        accent_color: dict_info.accent_color,
        toc_entries: &toc_entries,
        descriptions: &descriptions,
    };

    let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
    let html = TEMPLATES
        .render("word_group", &context)
        .map_err(|e| e.to_string())?;

    Ok((ids, html))
}

type DictData = (Vec<u8>, Vec<IdxEntry>);

fn create_dict(
    dict_info: &DictionaryInfo,
    words: impl Iterator<Item = impl PaliWord>,
    igen: &dyn InflectionGenerator,
    logger: &dyn EdpdLogger,
) -> Result<DictData, String> {
    logger.info(&"Creating dict entries.".to_string());
    let word_groups = words.group_by(|pw| pw.group_id());

    let mut dict_buffer: Vec<u8> = Vec::new();
    let mut idx_words: Vec<IdxEntry> = Vec::new();
    for (n, (key, word_group)) in (&word_groups).into_iter().enumerate() {
        let (ids, html_str) = get_ids_and_html_for_word_group(dict_info, word_group, igen, logger)?;

        let synonym_words: Vec<String> = ids
            .into_iter()
            .flat_map(|id| igen.generate_all_inflections(&id))
            .collect();
        let mut html_bytes = html_str.into_bytes();
        idx_words.push(IdxEntry {
            word: key,
            data_offset: if n == 0 {
                0
            } else {
                idx_words[n - 1].data_offset + idx_words[n - 1].data_size
            },
            data_size: html_bytes.len() as i32,
            synonym_words,
        });
        dict_buffer.append(&mut html_bytes);

        if n % 1_000 == 0 && n != 0 {
            logger.info(&format!(
                "... created {:05} dict entries, ending with '{}'.",
                n, idx_words[n].word
            ));
        }
    }

    logger.info(&format!(
        "... done creating {} dict entries.",
        idx_words.len()
    ));
    Ok((dict_buffer, idx_words))
}

fn create_idx(idx_entries: &[IdxEntry], logger: &dyn EdpdLogger) -> Vec<u8> {
    logger.info(&format!("Creating {} idx entries.", &idx_entries.len()));

    let idx: Vec<u8> = idx_entries.iter().fold(Vec::new(), |mut acc, e| {
        acc.append(&mut e.word.to_owned().into_bytes());
        acc.push(0u8);
        acc.extend_from_slice(&e.data_offset.to_be_bytes());
        acc.extend_from_slice(&e.data_size.to_be_bytes());
        acc
    });

    logger.info(&format!(
        "... done creating {} idx entries.",
        &idx_entries.len()
    ));
    idx
}

struct SynEntry {
    synonym_word: String,
    original_word_index: i32,
}

fn create_syn(idx_entries: &[IdxEntry], logger: &dyn EdpdLogger) -> (Vec<u8>, usize) {
    let mut syn_entries = idx_entries
        .iter()
        .enumerate()
        .fold(Vec::new(), |mut acc, (n, e)| {
            let mut ses: Vec<SynEntry> = e
                .synonym_words
                .iter()
                .map(|sw| SynEntry {
                    synonym_word: sw.to_owned(),
                    original_word_index: n as i32,
                })
                .collect();

            acc.append(&mut ses);

            acc
        });

    let syn_count = syn_entries.len();
    logger.info(&format!("Creating {} syn entries.", syn_count));

    syn_entries.sort_by(|w1, w2| glib::stardict_strcmp(&w1.synonym_word, &w2.synonym_word));
    let syn: Vec<u8> = syn_entries.iter().fold(Vec::new(), |mut acc, e| {
        acc.append(&mut e.synonym_word.to_owned().into_bytes());
        acc.push(0u8);
        acc.extend_from_slice(&e.original_word_index.to_be_bytes());
        acc
    });

    logger.info(&format!("... done creating {} syn entries.", syn_count));

    (syn, syn_count)
}

///
/// See https://github.com/huzheng001/stardict-3/blob/master/dict/doc/StarDictFileFormat
///
pub fn create_dictionary(
    dict_info: &DictionaryInfo,
    words: impl Iterator<Item = impl PaliWord>,
    igen: &dyn InflectionGenerator,
    logger: &dyn EdpdLogger,
) -> Result<Vec<StarDictFile>, String> {
    let (dict, mut idx_entries) = create_dict(dict_info, words, igen, logger)?;
    idx_entries.sort_by(|w1, w2| glib::stardict_strcmp(&w1.word, &w2.word));
    let idx = create_idx(&idx_entries, logger);
    let (syn, syn_count) = create_syn(&idx_entries, logger);
    let ifo = create_ifo(dict_info, idx_entries.len(), syn_count, idx.len())?;
    let png = create_png(dict_info);

    Ok(vec![
        StarDictFile {
            extension: "idx".to_string(),
            data: idx,
        },
        StarDictFile {
            extension: "dict".to_string(),
            data: dict,
        },
        StarDictFile {
            extension: "syn".to_string(),
            data: syn,
        },
        StarDictFile {
            extension: "ifo".to_string(),
            data: ifo,
        },
        StarDictFile {
            extension: "png".to_string(),
            data: png,
        },
    ])
}

fn create_ifo(
    dict_info: &DictionaryInfo,
    word_count: usize,
    syn_word_count: usize,
    idx_file_size: usize,
) -> Result<Vec<u8>, String> {
    let vm = IfoViewModel {
        name: dict_info.name,
        word_count,
        syn_word_count,
        idx_file_size,
        author: dict_info.author,
        description: dict_info.description,
        time_stamp: dict_info.time_stamp,
    };

    let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
    let ifo_str = TEMPLATES
        .render("ifo_file", &context)
        .map_err(|e| e.to_string())?;

    Ok(ifo_str.into_bytes())
}

fn create_png(dict_info: &DictionaryInfo) -> Vec<u8> {
    let mut png = Vec::new();
    png.extend_from_slice(&dict_info.ico);

    png
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tests::{TestInflectionGenerator, TestLogger};
    use crate::{resolve_file_in_manifest_dir, InputFormat, OutputFormat};

    #[derive(Debug, Deserialize)]
    struct TestPaliWord {
        id: String,
        sort_key: String,
        group_id: String,
        toc_id: String,
        toc_entry: String,
        word_data_entry: String,
    }

    impl PaliWord for TestPaliWord {
        fn id(&self) -> &str {
            &self.id
        }

        fn sort_key(&self) -> String {
            self.sort_key.clone()
        }

        fn group_id(&self) -> String {
            self.group_id.clone()
        }

        fn toc_id(&self, _input_format: &InputFormat) -> String {
            self.toc_id.clone()
        }

        fn toc_entry(&self, _input_format: &InputFormat) -> Result<String, String> {
            Ok(self.toc_entry.clone())
        }

        fn word_data_entry(
            &self,
            input_format: &InputFormat,
            feedback_form_url: &str,
            host_url: &str,
            host_version: &str,
            igen: &dyn InflectionGenerator,
        ) -> Result<String, String> {
            Ok(format!(
                "{}-{}-{}-{}-{}-{}",
                self.word_data_entry,
                input_format.to_string(),
                feedback_form_url,
                host_url,
                host_version,
                igen.generate_inflection_table_html(self.id())
            ))
        }
    }

    fn read_pali_words<'a>() -> impl Iterator<Item = impl PaliWord> + 'a {
        let path = resolve_file_in_manifest_dir(
            "src/stardict/output_generators/test_data/pali_words1.csv",
        )
        .expect("must exist!");

        let file = std::fs::File::open(path).expect("must exist");
        let rdr = csv::ReaderBuilder::new().from_reader(file);

        rdr.into_deserialize::<TestPaliWord>().map(|w| w.expect(""))
    }

    fn create_dict_info<'a>() -> DictionaryInfo<'a> {
        DictionaryInfo {
            name: "Digital Pāli Tools Dictionary (DPD)",
            input_format: &InputFormat::DevamittaPaliStudy,
            output_format: &OutputFormat::StarDict,
            author: "Digital Pāli Tools <digitalpalitools@gmail.com>",
            description: "The next generation comprehensive digital Pāli dictionary.",
            accent_color: "orange",
            time_stamp: "xxxx",
            ico: &[],
            feedback_form_url: "http://feedback.form/???",
            host_url: "this is the host",
            host_version: "host version",
            inflections_db_path: None,
        }
    }

    #[test]
    fn create_dict_test() {
        let words = read_pali_words();
        let igen = TestInflectionGenerator::new();

        let (dict_data, idx_entries) =
            create_dict(&create_dict_info(), words, &igen, &TestLogger::new()).expect("Unexpected");
        let dict_entries: Vec<String> = idx_entries
            .iter()
            .map(|ie| {
                let word_bytes =
                    &dict_data[ie.data_offset as usize..(ie.data_offset + ie.data_size) as usize];
                std::str::from_utf8(word_bytes)
                    .expect("unexpected")
                    .to_owned()
            })
            .collect();

        insta::assert_debug_snapshot!(dict_entries);
        insta::assert_yaml_snapshot!(idx_entries);
    }

    #[test]
    fn create_idx_test() {
        let idx_entries = vec![
            IdxEntry {
                word: "a".to_string(),
                data_offset: 1,
                data_size: 2,
                synonym_words: vec!["a1".to_string(), "a2".to_string()],
            },
            IdxEntry {
                word: "bc".to_string(),
                data_offset: 0x01000100,
                data_size: 0x00020002,
                synonym_words: vec!["b1".to_string(), "b2".to_string()],
            },
        ];

        let idx = create_idx(&idx_entries, &TestLogger::new());

        assert_eq!(
            idx,
            vec![0x61, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0x62, 0x63, 0, 1, 0, 1, 0, 0, 2, 0, 2]
        )
    }

    #[test]
    fn create_syn_test() {
        let idx_entries = vec![
            IdxEntry {
                word: "a".to_string(),
                data_offset: 1,
                data_size: 2,
                synonym_words: vec!["a1".to_string()],
            },
            IdxEntry {
                word: "bc".to_string(),
                data_offset: 0x01000100,
                data_size: 0x00020002,
                synonym_words: vec!["b1".to_string(), "b2".to_string()],
            },
        ];

        let (syn, syn_count) = create_syn(&idx_entries, &TestLogger::new());

        assert_eq!(
            syn,
            vec![0x61, 0x31, 0, 0, 0, 0, 0, 0x62, 0x31, 0, 0, 0, 0, 1, 0x62, 0x32, 0, 0, 0, 0, 1]
        );
        assert_eq!(syn_count, 3);
    }

    #[test]
    fn create_ifo_test() {
        let ifo = create_ifo(&create_dict_info(), 100, 500, 1000).expect("Unexpected");

        insta::assert_snapshot!(&String::from_utf8(ifo).expect("Unexpected"));
    }
}
