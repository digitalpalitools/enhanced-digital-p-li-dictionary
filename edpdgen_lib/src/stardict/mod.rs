use crate::inflection_generator::InflectionGenerator;
use crate::input::dpd::DpdPaliWord;
use crate::input::dps::DpsPaliWord;
use crate::input::input_format::InputFormat;
use crate::{DictionaryBuilder, DictionaryFile, DictionaryInfo, EdpdLogger};
use std::path::Path;

mod input_parsers;
mod output_generators;

pub struct StarDict<'a> {
    dict_info: &'a DictionaryInfo<'a>,
    input_data_path: &'a Path,
    igen: &'a dyn InflectionGenerator,
    logger: &'a dyn EdpdLogger,
}

impl<'a> DictionaryBuilder<'a> for StarDict<'a> {
    fn new(
        dict_info: &'a DictionaryInfo,
        input_data_path: &'a Path,
        igen: &'a dyn InflectionGenerator,
        logger: &'a dyn EdpdLogger,
    ) -> Self {
        StarDict {
            dict_info,
            input_data_path,
            igen,
            logger,
        }
    }

    fn build_files(&self) -> Result<Vec<DictionaryFile>, String> {
        match self.dict_info.input_format {
            InputFormat::Dpd => run_for_ods_type::<DpdPaliWord>(
                self.dict_info,
                self.input_data_path,
                self.igen,
                self.logger,
            ),
            InputFormat::Dps => run_for_ods_type::<DpsPaliWord>(
                self.dict_info,
                self.input_data_path,
                self.igen,
                self.logger,
            ),
        }
    }
}

pub trait StarDictPaliWord {
    fn id(&self) -> &str;
    fn sort_key(&self) -> String;
    fn group_id(&self) -> String;
    fn toc_id(&self, dict_short_name: &str) -> String;
    fn toc_entry(&self, dict_short_name: &str) -> Result<String, String>;
    fn word_data_entry(
        &self,
        dict_short_name: &str,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
    ) -> Result<String, String>;
}

pub fn run_for_ods_type<'a, T: 'a + serde::de::DeserializeOwned + StarDictPaliWord>(
    dict_info: &DictionaryInfo,
    input_data_path: &Path,
    igen: &dyn InflectionGenerator,
    logger: &dyn EdpdLogger,
) -> Result<Vec<DictionaryFile>, String> {
    let words = input_parsers::load_words::<T>(input_data_path, logger)?;
    let sd_files = output_generators::create_dictionary(&dict_info, words, igen, logger)?;

    Ok(sd_files)
}
