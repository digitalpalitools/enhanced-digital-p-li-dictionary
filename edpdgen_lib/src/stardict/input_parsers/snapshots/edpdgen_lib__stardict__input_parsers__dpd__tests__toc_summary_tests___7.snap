---
source: edpdgen_lib/src/stardict/input_parsers/dpd.rs
expression: toc_summary

---
<li><a class="dictionary-toc-link-dpy" href="#abalabala_dpy">abalabala</a>&nbsp;(adj)&nbsp;weak and strong</li>
